src/uiprotect/data/bootstrap.py

```diff
@@ -384,6 +386,58 @@ def _process_remove_packet(
             old_obj=device,
         )
 
+    def _process_ws_keyring_or_ulp_user_message(
+        self,
+        action: dict[str, Any],
+        data: dict[str, Any],
+        model_type: ModelType,
+    ) -> WSSubscriptionMessage | None:
+        action_id = action["id"]
+        dict_from_bootstrap: dict[str, ProtectModelWithId] = getattr(
+            self, to_snake_case(model_type.devices_key)
+        )
+        action_type = action["action"]
+        if action_type == "add":
+            add_obj = create_from_unifi_dict(data, api=self._api, model_type=model_type)
+            if TYPE_CHECKING:
+                model_class = MODEL_TO_CLASS.get(model_type)
+                assert model_class is not None and isinstance(add_obj, model_class)
+            add_obj = cast(ProtectModelWithId, add_obj)
+            dict_from_bootstrap[add_obj.id] = add_obj
+            return WSSubscriptionMessage(
+                action=WSAction.ADD,
+                new_update_id=self.last_update_id,
+                changed_data=add_obj.dict(),
+                new_obj=add_obj,
+            )
+        elif action_type == "remove":
+            removed_obj = dict_from_bootstrap.pop(action_id, None)
+            if removed_obj is None:
+                return None
+            return WSSubscriptionMessage(
+                action=WSAction.REMOVE,
+                new_update_id=self.last_update_id,
+                changed_data={},
+                old_obj=removed_obj,
+            )
+        elif action_type == "update":
+            updated_obj = dict_from_bootstrap.get(action_id)
+            if updated_obj is None:
+                return None
```

Comment:
Looks like there is missing coverage in the new code here

Reply:
resolved